File: stage2/sce_dx_study.cpp
// C++ Includes
#include <iostream>
#include <iomanip>
#include <fstream>
#include <unordered_map>
#include <cmath>

// ROOT Includes
#include "TROOT.h"
#include <TStyle.h>
#include "ROOT/TThreadedObject.hxx"
#include <ROOT/RDataFrame.hxx>
#include "TTree.h"
#include "TFile.h"
#include "TSystemFile.h"
#include "TString.h"
#include "TF1.h"
#include "TH1.h"
#include "TH2.h"
#include "TTreeReader.h"
#include "TTreeReaderArray.h"
#include "TVirtualFFT.h"
#include "TFile.h"
#include "TChain.h"
#include "TSystemDirectory.h"

using namespace std;

// Function Declaration
vector<double> linspace(double start, double end, int n);

//////////////////////////////////////////////////
// Main Analysis
//////////////////////////////////////////////////
int main(int argc, char** argv)
{
  // Drift Velocity vs. E-Field fit
  TF1 driftVelFit("driftVelFit","(1.0 - 0.0184 * (87.5 - 89.0)) * pol5",-0.05,1.1);
  driftVelFit.SetParameter(0,0.0);
  driftVelFit.SetParameter(1,5.53416);
  driftVelFit.SetParameter(2,-6.53093);
  driftVelFit.SetParameter(3,3.20752);
  driftVelFit.SetParameter(4,0.389696);
  driftVelFit.SetParameter(5,-0.556184);

  // Load data
  Char_t *inputfilename = (Char_t*)"";
  inputfilename = argv[1];
  unique_ptr<TFile> infile( TFile::Open(inputfilename) );
  unique_ptr<TH3> hAgg(infile->Get<TH3>("DummyHist3D"));
  
  char *tpc;
  float xrange0;
  float xrange1;
  tpc = (char*) argv[2];
  int sel_tpc;
  if (!strcmp(tpc, "EE"))
    {
      xrange0=-358.49;
      xrange1=-210.29;
      sel_tpc = 0;
    }
  else if (!strcmp(tpc, "EW"))
    {
      xrange0=-210.14;
      xrange1=-61.94;
      sel_tpc = 1;
    }
  else if (!strcmp(tpc, "WE"))
    {
      xrange0=61.94;
      xrange1=210.14;
      sel_tpc = 2;
    }
  else if(!strcmp(tpc, "WW"))
    {
      xrange0=210.29;
      xrange1=358.49;
      sel_tpc = 3;
    }

  // Output file
  string scedxstudy_text = string("scedxstudy_") + tpc + string(".root");
  TFile outf(scedxstudy_text.c_str(), "recreate");
  outf.cd();

  // Make output TH3s, starting from a copy of input TH3
  TH3D *hEfield3D = new TH3D("hEfield3D","",hAgg->GetNbinsX(), xrange0, xrange1, hAgg->GetNbinsY(), -181.86, 134.96, hAgg->GetNbinsZ(), -894.9515, 894.9515);
  TH3D *hRho3D = new TH3D("hRho3D","",hAgg->GetNbinsX(), xrange0, xrange1, hAgg->GetNbinsY(), -181.86, 134.96, hAgg->GetNbinsZ(), -894.9515, 894.9515);
  TH3D *hRho3D_clean = new TH3D("hRho3D_clean","",hAgg->GetNbinsX(), xrange0, xrange1, hAgg->GetNbinsY(), -181.86, 134.96, hAgg->GetNbinsZ(), -894.9515, 894.9515);

  // Loop over {y,z} bins
  for (int j = 0; j <= hAgg->GetNbinsY()+1; j++)
    {
      float bin_low_y = hAgg->GetYaxis()->GetBinLowEdge(j);
      float bin_high_y = hAgg->GetYaxis()->GetBinUpEdge(j);

      for (int k = 0; k <= hAgg->GetNbinsZ()+1; k++)
        {
          float bin_low_z = hAgg->GetZaxis()->GetBinLowEdge(k);
          float bin_high_z = hAgg->GetZaxis()->GetBinUpEdge(k);

	  // Create unique (output) directory for every {y,z} bin
	  string dirstring = to_string(j) + "_" + to_string(k);
          TDirectory* yzbindir = outf.mkdir(dirstring.c_str());
          yzbindir->cd();

	  ////////////////////////////////////////////////////////////
	  /// Spatial Offsets
	  ////////////////////////////////////////////////////////////

	  // "Project" a single {y,,z} bin along x to get dx vs. x histo
	  TH1D *hdx = hAgg->ProjectionX("hdx", j, j, k, k, "o");

	  // Convert hdx to gdx (TGraph)
	  vector<double> gx;
	  vector<double> gy;
	  for (int i = 1; i <= hdx->GetNbinsX(); i++)
	    {
	      gx.push_back(hdx->GetXaxis()->GetBinCenter(i));
	      gy.push_back(hdx->GetBinContent(i));
	    }

	  TGraph *gdx = new TGraph(gx.size(), gx.data(), gy.data());
	  gdx->SetName("gdx");
	  TF1 *dxfit = new TF1("dxfit","pol3",xrange0,xrange1); // fit
	  gdx->Fit("dxfit", "MRQFB");
	  gdx->SetTitle("Spatial Offsets vs. Drift Length");
          gdx->GetXaxis()->SetTitle("Drift Length [cm]");
          gdx->GetYaxis()->SetTitle("Spatial Offset dX [cm]");
	  gdx->Write();
	  
	  /////////////////////////////////////////////////////////
	  /// Local Drift Velocity
	  /////////////////////////////////////////////////////////
	  
	  // Sample dx vs. x fit on a linspace
	  vector<double> tempx = linspace(xrange0, xrange1, 100);
          vector<double> tempy;
          for(const auto& x : tempx)
            {
              tempy.push_back(dxfit->Eval(x));
            }
	  
	  // Calculate difference between adjacent dxs
	  vector<double> ddxs;
          adjacent_difference(tempy.begin(), tempy.end(), back_inserter(ddxs));
          ddxs.erase(ddxs.begin());
	  
	  // Velocity calculation
	  vector<double> loc_v;
	  double spacing = tempx[1] - tempx[0];
          for(const auto& ddx : ddxs)
            {
              float vel = 0.157565 * (1 - ddx / spacing);
              loc_v.push_back(vel);
            }
	  
	  vector<double> xadj;
          for(const auto& x : tempx)
            {
              xadj.push_back(x + spacing/2);
            }
          xadj.pop_back();

	  TGraph *gVel = new TGraph(xadj.size(), xadj.data(), loc_v.data());
          gVel->SetName("gVel");
          gVel->SetTitle("Drift Velocity vs. Drift Length");
	  gVel->GetXaxis()->SetTitle("Drift Length [cm]");
	  gVel->GetYaxis()->SetTitle("Drift Velocity [cm/#mus]");
	  gVel->Write();

	  /////////////////////////////////////////////////////////////
	  /// Local Electric Field
	  /////////////////////////////////////////////////////////////
	  
	  // Get E-field from known relationship with drift velocity
	  vector<double> loc_E;
          for(const auto& v : loc_v)
            {
              loc_E.push_back(driftVelFit.GetX(10*v));
            }

	  // E-field vs. drift length
	  TGraph *gEfield = new TGraph(xadj.size(), xadj.data(), loc_E.data());
          gEfield->SetName("gEfield");
	  gEfield->SetTitle("Electric Field vs. Drift Length");
	  gEfield->GetXaxis()->SetTitle("Drift Length [cm]");
	  gEfield->GetYaxis()->SetTitle("Electric Field [kV/cm]");
          //TF1 *efit = new TF1("efit","[0]*exp([1]*(x-[2])) + [3]",xrange0,xrange1);
	  //efit->SetParameters(10,0.1,200,0.48);								      
	  //TF1 *efit = new TF1("efit","[0]*(x-60)**3 + [1]*(x-60)**2 + [2]*(x-60)**1 + [3]*(x-60)**0",xrange0,xrange1);
	  //efit->SetParameters(0.000000001,0.000001,0.0001,0.48);
	  //TF1 *efit0 = new TF1("efit0","pol2",xrange0,xrange0+30);
	  TF1 *efit = new TF1("efit","pol3",xrange0,xrange1);
	  //TF1 *efit = new TF1("efit","efit0 + efit1",xrange0,xrange1);
	  //TF1 *efit = new TF1("efit","[0]*[1]**(x-60) + 0.48",xrange0,xrange1);
	  //efit->SetParameters(0.00001,1.08);
          gEfield->Fit("efit", "MRQF");
	  gEfield->Write();

	  ///////////////////////////////////////////////////////////////////
	  /// Local Space Charge Density
	  ///////////////////////////////////////////////////////////////////

	  // Sample dx vs. x fit on a linspace
	  vector<double> efieldvals;
	  vector<double> rhovals;
          for(const auto& x : tempx)
            {
	      efieldvals.push_back(efit->Eval(x)); // Doing this now to use for TH3 later
              rhovals.push_back((pow(10.0,9))*0.00013316416*(efit->Derivative(x)));
            }	  

	  // Space charge density vs. drift length
          TGraph *gRho = new TGraph(tempx.size(), tempx.data(), rhovals.data());
          gRho->SetName("gRho");
	  gRho->SetTitle("Space Charge Density vs. Drift Length");
	  gRho->GetXaxis()->SetTitle("Drift Length [cm]");
	  gRho->GetYaxis()->SetTitle("Space Charge Density [nC/m^{3}]");
	  gRho->Write();

	  ///////////////////////////////////////////////////////////////////
	  /// Create Efield and SC density maps
	  ///////////////////////////////////////////////////////////////////

	  // Average over bins of x
	  for (int i = 0; i <= hAgg->GetNbinsX()+1; i++)
            {
              float bin_low_x = hAgg->GetXaxis()->GetBinLowEdge(i);
              float bin_high_x = hAgg->GetXaxis()->GetBinUpEdge(i);

              int gbn = hAgg->GetBin(i, j, k);
	      
	      // Sum entries for specific x-bin
	      vector<double> xbin_efieldvals;
              vector<double> xbin_rhovals;
              for(int s=0; s<efieldvals.size(); s++)
                {
                  if(tempx[s] > bin_low_x && tempx[s] < bin_high_x)
                    {
		      xbin_efieldvals.push_back(efieldvals[s]);
                      xbin_rhovals.push_back(rhovals[s]);
                    }
                }

	      // Average caculation
	      double efield_sum = 0;
	      for(const auto& val : xbin_efieldvals)
		{
		  efield_sum += val;
		}
	      double xbin_efieldmean = efield_sum / xbin_efieldvals.size();
	      hEfield3D->SetBinContent(gbn, xbin_efieldmean);

              double rho_sum = 0;
              for(const auto& val : xbin_rhovals)
		{
                  rho_sum += val;
                }
	      double xbin_rhomean = rho_sum / xbin_rhovals.size();
	      hRho3D->SetBinContent(gbn, xbin_rhomean);
	      
	    } // end x loop
	} // end z loop
    } // end y loop
  

  /////////////////////////////////////////////////////////////
  /// Clean up ... Using hRho3D, fill ONLY good bins of new TH3
  //////////////////////////////////////////////////////////////
  
  // Make lines (y=mx+b) defining y/z cuts  
  for (int i = 0; i <= hAgg->GetNbinsX()+1; i++)
    {
      float bin_cen_x = hAgg->GetXaxis()->GetBinCenter(i);
      for (int j = 0; j <= hAgg->GetNbinsY()+1; j++)
	{
	  float bin_cen_y = hAgg->GetYaxis()->GetBinCenter(j);
	  for (int k = 0; k <= hAgg->GetNbinsY()+1; k++)
	    {
	      float bin_cen_z = hAgg->GetZaxis()->GetBinLowEdge(k);
	      int gbn = hAgg->GetBin(i, j, k);

	      // exclude nasty bins
	      float top_y_cut = -0.46*bin_cen_x + 120;
	      float bottom_y_cut = 0.46*bin_cen_x -166;
	      if(bin_cen_y < top_y_cut && bin_cen_y > bottom_y_cut && bin_cen_z > -800 && bin_cen_z < 800)
		{
		  hRho3D_clean->SetBinContent(gbn, hRho3D->GetBinContent(gbn));
		}
	    }
	}
    }
  

  outf.cd();
  
  hEfield3D->GetXaxis()->SetTitle("x [cm]");
  hEfield3D->GetYaxis()->SetTitle("y [cm]");
  hEfield3D->GetZaxis()->SetTitle("z [cm]");
  hEfield3D->Write();

  hRho3D->GetXaxis()->SetTitle("x [cm]");
  hRho3D->GetYaxis()->SetTitle("y [cm]");
  hRho3D->GetZaxis()->SetTitle("z [cm]");
  hRho3D->Write();

  TH2D *hRhoXY = (TH2D*) hRho3D->Project3D("yx");
  hRhoXY->SetTitle("Average Space Charge Density [nC/m^{3}]");
  hRhoXY->GetXaxis()->SetTitle("x [cm]");
  hRhoXY->GetYaxis()->SetTitle("y [cm]");
  hRhoXY->Write();

  hRho3D->GetZaxis()->SetRange(5,5);
  TH2D *hRhoXY_Z0 = (TH2D*) hRho3D->Project3D("Z0_yx");
  hRhoXY_Z0->SetTitle("Average Space Charge Density [nC/m^{3}]: Z = 0 cm");
  hRhoXY_Z0->GetXaxis()->SetTitle("x [cm]");
  hRhoXY_Z0->GetYaxis()->SetTitle("y [cm]");
  hRhoXY_Z0->Write();

  
  hRho3D_clean->GetXaxis()->SetTitle("x [cm]");
  hRho3D_clean->GetYaxis()->SetTitle("y [cm]");
  hRho3D_clean->GetZaxis()->SetTitle("z [cm]");
  hRho3D_clean->Write();
  TH2D *hRhoXY_clean = (TH2D*) hRho3D_clean->Project3D("yx");
  hRhoXY_clean->SetTitle("Average Space Charge Density [nC/m^{3}]");
  hRhoXY_clean->GetXaxis()->SetTitle("x [cm]");
  hRhoXY_clean->GetYaxis()->SetTitle("y [cm]");
  hRhoXY_clean->Write();
  outf.Close();
  
  return 0;

} // end main

/////////////////////////////////////////////////
/// Functions
/////////////////////////////////////////////////

// Create linspace, assume n > 1
vector<double> linspace(double start, double end, int n)
{
  // output
  vector<double> xs;
  
  double delta = (end - start) / (n - 1);
  for(int i=0; i < n-1; ++i)
    {
      xs.push_back(start + delta * i);
    }

  xs.push_back(end);
  return xs;
}

File: stage2/sce_dx_study.py
# imports
import ROOT
import numpy as np
import pandas as pd
import argparse

def main(args):
    
    # Load/prepare histograms
    outfile = ROOT.TFile.Open('teststudy.root', 'RECREATE')
    histfile = ROOT.TFile.Open(args.in_file_hist, 'READ')
    dummyhist = histfile.Get('DummyHist3D')
    nbins_x = dummyhist.GetNbinsX()
    nbins_y = dummyhist.GetNbinsY()
    nbins_z = dummyhist.GetNbinsZ()
    hreco = dummyhist.Clone('hreco')
    hsim = dummyhist.Clone('hsim')

    # Load reco/sim aggregated offsets into pandas dataframes
    dfs = []
    for arg in vars(args):
        if (arg == 'in_file_reco' or arg == 'in_file_sim') and (getattr(args, arg)):
            df = pd.read_csv(getattr(args,arg))
            df['gbn'] = df['gbn'].astype(int)
            dfs.append(df)

    # Fill TH3s with aggregated offsets
    for i in range(dummyhist.GetNbinsX() + 2):
        for j in range(dummyhist.GetNbinsY() + 2):
            for k in range(dummyhist.GetNbinsZ() + 2):
                gbn = dummyhist.GetBin(i, j, k)
                # Reco
                if gbn in dfs[0]['gbn'].values:
                    hreco.SetBinContent(gbn, dfs[0][dfs[0]['gbn'] == gbn]['med'].values[0])
                # Sim
                if (len(dfs) > 1) and (gbn in dfs[1]['gbn'].values):
                    hsim.SetBinContent(gbn, dfs[1][dfs[1]['gbn'] == gbn]['med'].values[0])

    # Analysis
    outfile.cd()
    print(hreco.GetNbinsZ())
    for j in range(dummyhist.GetNbinsY() + 2):
        for k in range(dummyhist.GetNbinsZ() + 2):
            
            print(j,k)

            # We are now working in a distinct y-z bin
            # Create a directory for this bin
            dirstring = str(j) + "_" + str(k)
            yzbindir = outfile.mkdir(dirstring)
            yzbindir.cd()

            ##################################################
            ### dx vs x
            ##################################################
            # reco
            hdx = hreco.ProjectionX('hdx', j, j, k, k, 'o')
            gx, gx_low, gx_high = [], [], []
            gy, gy_low, gy_high = [], [], []
            for i in range(hreco.GetNbinsX() + 2):
                gbn = hreco.FindBin(hdx.GetXaxis().GetBinCenter(i), hreco.GetYaxis().GetBinCenter(j), hreco.GetZaxis().GetBinCenter(k))
                gx.append(hdx.GetXaxis().GetBinCenter(i))
                gx_low.append(hdx.GetXaxis().GetBinWidth(i)/2)
                gx_high.append(hdx.GetXaxis().GetBinWidth(i)/2)
                gy.append(hdx.GetBinContent(i))
                if gbn in dfs[0].gbn.values:
                    gy_low.append(hdx.GetBinContent(i) - dfs[0][dfs[0]['gbn'] == gbn]['med_low'].values[0])
                    gy_high.append(dfs[0][dfs[0]['gbn'] == gbn]['med_high'].values[0] - hdx.GetBinContent(i))
                else:
                    gy_low.append(hdx.GetBinContent(i))
                    gy_high.append(hdx.GetBinContent(i))

            gdx = ROOT.TGraphAsymmErrors(len(gx), np.array(gx, dtype='float'), np.array(gy, dtype='float'), np.array(gx_low, dtype='float'), np.array(gx_high, dtype='float'), np.array(gy_low, dtype='float'), np.array(gy_high, dtype='float'))
            gdx.SetName('gdx')
            dxfit = ROOT.TF1('dxfit','pol3',hreco.GetXaxis().GetBinLowEdge(1), hreco.GetXaxis().GetBinLowEdge(hreco.GetNbinsX() + 1)); # fit function
            gdx.Fit('dxfit', 'MRQFEX0')
            gdx.Write()

            # sim
            if len(dfs) > 1:
                hdx_sim = hsim.ProjectionX('hdx_sim', j, j, k, k, 'o')
                gx_sim, gy_sim = [], []
                for i in range(hsim.GetNbinsX() + 2):
                    gbn = hsim.FindBin(hdx_sim.GetXaxis().GetBinCenter(i), hsim.GetYaxis().GetBinCenter(j), hsim.GetZaxis().GetBinCenter(k))
                    gx_sim.append(hdx_sim.GetXaxis().GetBinCenter(i))
                    gy_sim.append(hdx_sim.GetBinContent(i))
                gdx_sim = ROOT.TGraph(len(gx_sim), np.array(gx_sim, dtype='float'), np.array(gy_sim, dtype='float'))
                gdx_sim.SetName('gdx_sim')
                gdx_sim.Write()
               

            ##################################################
            ### Local drift velocity
            ##################################################
            tempx = np.linspace(hreco.GetXaxis().GetBinLowEdge(1), hreco.GetXaxis().GetBinLowEdge(hreco.GetNbinsX() + 1), 100)
            tempy = [dxfit.Eval(x) for x in tempx]
            
            # Calculate difference between adjacent dxs
            ddxs = np.diff(tempy)
            spacing = np.diff(tempx)[0]
            loc_v = [0.157565 * (1 - ddx / spacing) for ddx in ddxs]
            xadj = [x + spacing/2 for x in tempx]
            del xadj[-1]
            
            gvel = ROOT.TGraph(len(xadj), np.array(xadj, dtype='float'), np.array(loc_v, dtype='float'))
            gvel.SetName('gvel')
            gvel.Write()

            ##################################################
            ### Local efield
            ##################################################
            # Fit for efield using know relationship with drift velocity
            driftvelfit = ROOT.TF1('driftvelfit','(1.0 - 0.0184 * (87.5 - 89.0)) * pol5', -0.05, 1.1)
            driftvelfit.SetParameter(0, 0.0)
            driftvelfit.SetParameter(1, 5.53416)
            driftvelfit.SetParameter(2, -6.53093)
            driftvelfit.SetParameter(3, 3.20752)
            driftvelfit.SetParameter(4, 0.389696)
            driftvelfit.SetParameter(5, -0.556184)
            loc_e = [driftvelfit.GetX(10*v) for v in loc_v]
            
            # Efield vs. x
            gefield = ROOT.TGraph(len(xadj), np.array(xadj, dtype='float'), np.array(loc_e, dtype='float'))
            # Fit curve to efield, so that we can take its derivative to find rho
            efit = ROOT.TF1('efit', 'pol3', hreco.GetXaxis().GetBinLowEdge(1), hreco.GetXaxis().GetBinLowEdge(hreco.GetNbinsX() + 1))
            gefield.Fit("efit", "MRQF");
            gefield.SetName('gefield')
            gefield.Write()

            ##################################################
            ### Local space charge density
            ##################################################
            rho = [10**9 * 0.00013316416 * efit.Derivative(x) for x in xadj]
            
            # rho vs. x
            grho = ROOT.TGraph(len(xadj), np.array(xadj, dtype='float'), np.array(rho, dtype='float'))
            grho.SetName('grho')
            grho.Write()

            
    
    # Plot
    yz_bin = '6_5'
    gdxs = [outfile.Get(yz_bin + '/gdx'), outfile.Get(yz_bin + '/gdx_sim')]
    plot_yz_bin_dx_vs_x(gdxs)

    histfile.Close()
    outfile.Close()




def plot_yz_bin_dx_vs_x(gdxs):
    c = ROOT.TCanvas('c', '', 800, 800)
    c.cd()

    c.SetGrid();
    #mg = ROOT.TMultiGraph('mg', '')
    
    # reco
    g0 = gdxs[0]
    g0.GetFunction('dxfit').Delete()
    g0.SetTitle('Reco')
    g0.SetMarkerStyle(21)
    g0.SetMarkerColor(39)
    g0.GetHistogram().SetMinimum(-0.1)
    g0.GetHistogram().SetMaximum(1.0)
    g0.Draw('AP')
    #mg.Add(g0)
    
    legend = ROOT.TLegend(0.55,0.65,0.76,0.82);
    legend.SetTextSize(0.03)
    legend.AddEntry(g0, 'Reco', 'lp')

    # sim
    if len(gdxs) > 1:
        g1 = gdxs[1]
        g1.SetTitle('Simulation')
        g1.SetDrawOption('L')
        g1.Draw('SAME')
        legend.AddEntry(g1, 'Sim', 'lp')

    legend.Draw()

    c.SaveAs('test.png')            


  
if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--in_file_hist', type=str, required=True)
    parser.add_argument('--in_file_reco', type=str, required=True)
    parser.add_argument('--in_file_sim', type=str, required=False)
    parser.add_argument('--yz_bin', type=str, required=False)
    args = parser.parse_args()
    main(args)
